### PR TITLE
Upgrade gRPC

### PR DESCRIPTION
**Pull request by cocreature**
The main reason for this upgrade is that it might fix some segfaults
in damlc that came from the gRPC bindings but I’ve also bumped the
JVM side at the same time.

The only slightly non-trivial change is that grpc-java stopped
exporting the compiler plugin since you are now supposed to use the
high-level rules/the toolchain. However, that doesn’t quite work for
the ledger bindings since they don’t have the provider needed for
javadoc so I patched that for now (by just changing the
visibility). I’ll open an issue for that in grpc-java.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: deps.bzl [hunk 1]
```diff
@@ -164,9 +164,16 @@ def daml_deps():
...
             name = "io_grpc_grpc_java",
-            strip_prefix = "grpc-java-1.19.0",
-            urls = ["https://github.com/grpc/grpc-java/archive/v1.19.0.tar.gz"],
-            sha256 = "81d1e12bf0f8bd1560eed7c75f24d8bb8e7368dcf07802586e439c85cf89b005",
+            strip_prefix = "grpc-java-7e5e6e604a9a74cb62a0aa6edbdb47f2ae676ed7",
+            # This corresponds to the v1.21.x branch which is slightly ahead of the latest release
+            # in that series and has a fix for using custom Java toolchains.
+            urls = ["https://github.com/grpc/grpc-java/archive/7e5e6e604a9a74cb62a0aa6edbdb47f2ae676ed7.tar.gz"],
+            sha256 = "92dea47bdb47929465968b564238f9f8d8fcd28765e4c446380a819ca6fb5208",
+            # The java_grpc_library rules do not produce a "java" provider
```
Comment (aherrmann-da) [inline deps.bzl]: Is the issue a missing `xyz.java` or a missing `xyz[Java]`? I think the former is being deprecated, maybe the latter is present?

Comment (cocreature) [reply deps.bzl]: It complains about the former. It looks like it provides a `JavaInfo` provider so I guess `bazel_common` needs to switch over their `javadoc_library` to use that.

**Comment by cocreature** [comment]
It looks like we need to upgrade the MacOS frameworks in nix if we want to make this work. This has already be done in nixpkgs-unstable https://github.com/NixOS/nixpkgs/pull/56744/files but the diff looks quite large so I’m somewhat scared to attempt to backport this. Switching to nixpkgs-unstable is probably the easier option but that also comes with its own set of issues.

**Comment by neil-da** [comment]
What do we need from an upgraded Nix framework? Any chance we can pull those dependencies in via Bazel instead? Stopping segfaults and upgrading grpc both seem very valuable.

**Comment by cocreature** [comment]
We don’t need an upgraded nix framework, we need the MacOS frameworks that we currently get from nix. These are tied fairly closely to our CC toolchain so I don’t think we can pull them from Bazel. One option might be to rely on installing the MacOS frameworks separately via xcode instead of getting them from nix.

I do agree that we need to do something about this. At the moment, the easiest option seems to be to switch to `nixpkgs-unstable` which given that we don’t get that much stuff from nixpkgs might not be that unreasonable (fwiw I run nixos-unstable on my personal laptop and it’s a lot more stable than the name might suggest :slightly_smiling_face:).

**Comment by neil-da** [comment]
I'm perfectly happy with using nix unstable. I'm also happy with imposing an XCode requirement. In both cases we are gradually cutting down our Nix footprint, which was the semi-plan all along.

**Comment by cocreature** [comment]
Closing for now, I’m working on switching to over to nixpkgs-unstable and will open a new PR on top of that.

**Pull request closed without merging**